src/Phobos.Kafka/src/Petabridge.Phobos.Kafka.Producer/PhobosActorRefSourceActor.cs

```diff
@@ -0,0 +1,205 @@
+        /// <param name="maxFixedBufferSize">TBD</param>
+        public PhobosActorRefSourceActor(int bufferSize, OverflowStrategy overflowStrategy, int maxFixedBufferSize)
+        {
+            BufferSize = bufferSize;
+            OverflowStrategy = overflowStrategy;
+            Buffer = bufferSize != 0 ?  Producer.Buffer.Create<(T, ISpanContext)>(bufferSize, maxFixedBufferSize) : null;
+        }
+
+        /// <summary>
+        /// TBD
+        /// </summary>
+        protected ILoggingAdapter Log => _log ??= Context.GetLogger();
+
+        /// <summary>
+        /// TBD
+        /// </summary>
+        /// <param name="message">TBD</param>
+        /// <returns>TBD</returns>
+        protected override bool Receive(object message)
+            => DefaultReceive(message) || RequestElement(message) || (message is T && ReceiveElement((T) message));
+
+        /// <summary>
+        /// TBD
+        /// </summary>
+        /// <param name="message">TBD</param>
+        /// <returns>TBD</returns>
+        protected bool DefaultReceive(object message)
+        {
+            if (message is Cancel)
+                Context.Stop(Self);
+            else if (message is Status.Success)
+            {
+                if (BufferSize == 0 || Buffer.IsEmpty)
+                    Context.Stop(Self);  // will complete the stream successfully
+                else
+                    Context.Become(DrainBufferThenComplete);
+            }
+            else if (message is Status.Failure && IsActive)
+                OnErrorThenStop(((Status.Failure)message).Cause);
+            else
+                return false;
+            return true;
+        }
+
+        /// <summary>
+        /// TBD
+        /// </summary>
+        /// <param name="message">TBD</param>
+        /// <returns>TBD</returns>
+        protected virtual bool RequestElement(object message)
+        {
+            if (message is Request)
+            {
+                // totalDemand is tracked by base
+                if (BufferSize != 0)
+                    while (TotalDemand > 0L && !Buffer.IsEmpty)
+                        OnNext(Buffer.Dequeue());
+
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// TBD
+        /// </summary>
+        /// <param name="element">TBD</param>
+        /// <returns>TBD</returns>
+        protected virtual bool ReceiveElement(T element)
+        {
+            if (IsActive)
+            {
+                var context = Context.GetInstrumentation().ActiveSpan?.Context;
+                var message = (element, context);
+                
+                if (TotalDemand > 0L)
+                    OnNext(message);
+                else if (BufferSize == 0)
+                    Log.Debug("Dropping element because there is no downstream demand: [{0}]", message);
+                else if (!Buffer.IsFull)
+                    Buffer.Enqueue(message);
```

Comment:
The main reason why I think this is the best way of doing this, we don't know if the span context is stale or not, since it might be buffered by the ActorRef source. We probably also need a special actor for the publisher actor.